### PR TITLE
sql: improve EXPLAIN (DDL) output

### PR DESCRIPTION
**Pull request by postamar**
Previously, the EXPLAIN (DDL, STAGES) and EXPLAIN (DDL, DEPS) statements
were printing graphviz URLs to help engineers during the development of
the declarative schema changer.

Now that the declarative schema changer is going live, the audience for
EXPLAIN (DDL) changes to become the users. The graphviz URLs were, at
any rate, not very useful for schema changes involving many objects, due
to limitations in the rendering engine.

This commit gets rid of the STAGES and DEPS flags to provide a unified
plan which is similar to the data-driven test output in the scplan
package. The VERBOSE flag can now be used to render a superset of the
information in the default rendering.

Touches #77158.

Release note (sql change): EXPLAIN (DDL), when invoked on statements
supported by the declarative schema changer, prints a plan of what the
schema changer will do. This can be useful for the user to anticipate
the complexity of a schema change (anything involving Backfill or
Validation operations might be slow to run). This can be useful for
troubleshooting. EXPLAIN (DDL, VERBOSE) produces a more detailed plan.

-----

`EXPLAIN (DDL)` Example:
```
Schema change plan for DROP VIEW ‹test›.public.‹v1dep› CASCADE;
 ├── StatementPhase
 │    └── Stage 1 of 1 in StatementPhase
 │         ├── 4 elements transitioning toward ABSENT
 │         │    ├── PUBLIC → TXN_DROPPED View:{DescID: 147}
 │         │    ├── PUBLIC → TXN_DROPPED View:{DescID: 148}
 │         │    ├── PUBLIC → TXN_DROPPED View:{DescID: 149}
 │         │    └── PUBLIC → TXN_DROPPED View:{DescID: 150}
 │         └── 4 Mutation operations
 │              ├── MarkDescriptorAsDroppedSynthetically {"DescID":147}
 │              ├── MarkDescriptorAsDroppedSynthetically {"DescID":148}
 │              ├── MarkDescriptorAsDroppedSynthetically {"DescID":149}
 │              └── MarkDescriptorAsDroppedSynthetically {"DescID":150}
 ├── PreCommitPhase
 │    └── Stage 1 of 1 in PreCommitPhase
 │         ├── 56 elements transitioning toward ABSENT
 │         │    ├──      PUBLIC → ABSENT     Namespace:{DescID: 147, Name: v1dep, ReferencedDescID: 104}
 │         │    ├──      PUBLIC → ABSENT     Owner:{DescID: 147}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 147, Name: admin}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 147, Name: root}
 │         │    ├── TXN_DROPPED → DROPPED    View:{DescID: 147}
 │         │    ├──      PUBLIC → ABSENT     ObjectParent:{DescID: 147, ReferencedDescID: 105}
 │         │    ├──      PUBLIC → ABSENT     TableComment:{DescID: 147}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 147, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 147, Name: name, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 147, ColumnFamilyID: 0, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 147, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     Namespace:{DescID: 148, Name: v2dep, ReferencedDescID: 104}
 │         │    ├──      PUBLIC → ABSENT     Owner:{DescID: 148}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 148, Name: admin}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 148, Name: root}
 │         │    ├── TXN_DROPPED → DROPPED    View:{DescID: 148}
 │         │    ├──      PUBLIC → ABSENT     ObjectParent:{DescID: 148, ReferencedDescID: 105}
 │         │    ├──      PUBLIC → ABSENT     TableComment:{DescID: 148}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 148, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 148, Name: n1, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 148, ColumnFamilyID: 0, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 148, ColumnID: 1}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 148, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 148, Name: n2, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 148, ColumnFamilyID: 0, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 148, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     Namespace:{DescID: 149, Name: v3dep, ReferencedDescID: 104}
 │         │    ├──      PUBLIC → ABSENT     Owner:{DescID: 149}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 149, Name: admin}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 149, Name: root}
 │         │    ├── TXN_DROPPED → DROPPED    View:{DescID: 149}
 │         │    ├──      PUBLIC → ABSENT     ObjectParent:{DescID: 149, ReferencedDescID: 105}
 │         │    ├──      PUBLIC → ABSENT     TableComment:{DescID: 149}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 149, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 149, Name: name, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 149, ColumnFamilyID: 0, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 149, ColumnID: 1}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 149, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 149, Name: n1, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 149, ColumnFamilyID: 0, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 149, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     Namespace:{DescID: 150, Name: v4dep, ReferencedDescID: 104}
 │         │    ├──      PUBLIC → ABSENT     Owner:{DescID: 150}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 150, Name: admin}
 │         │    ├──      PUBLIC → ABSENT     UserPrivileges:{DescID: 150, Name: root}
 │         │    ├── TXN_DROPPED → DROPPED    View:{DescID: 150}
 │         │    ├──      PUBLIC → ABSENT     ObjectParent:{DescID: 150, ReferencedDescID: 105}
 │         │    ├──      PUBLIC → ABSENT     TableComment:{DescID: 150}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 150, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 150, Name: n2, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 150, ColumnFamilyID: 0, ColumnID: 1}
 │         │    ├──      PUBLIC → ABSENT     ColumnComment:{DescID: 150, ColumnID: 1}
 │         │    ├──      PUBLIC → WRITE_ONLY Column:{DescID: 150, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnName:{DescID: 150, Name: n1, ColumnID: 2}
 │         │    ├──      PUBLIC → ABSENT     ColumnType:{DescID: 150, ColumnFamilyID: 0, ColumnID: 2}
 │         │    └──      PUBLIC → ABSENT     ColumnComment:{DescID: 150, ColumnID: 2}
 │         └── 29 Mutation operations
 │              ├── DrainDescriptorName {"Namespace":{"DatabaseID":104,"DescriptorID":147,"Name":"v1dep","SchemaID":105}}
 │              ├── RemoveDroppedColumnType {"ColumnID":1,"TableID":147}
 │              ├── DrainDescriptorName {"Namespace":{"DatabaseID":104,"DescriptorID":148,"Name":"v2dep","SchemaID":105}}
 │              ├── RemoveDroppedColumnType {"ColumnID":1,"TableID":148}
 │              ├── RemoveDroppedColumnType {"ColumnID":2,"TableID":148}
 │              ├── DrainDescriptorName {"Namespace":{"DatabaseID":104,"DescriptorID":149,"Name":"v3dep","SchemaID":105}}
 │              ├── RemoveDroppedColumnType {"ColumnID":1,"TableID":149}
 │              ├── RemoveDroppedColumnType {"ColumnID":2,"TableID":149}
 │              ├── DrainDescriptorName {"Namespace":{"DatabaseID":104,"DescriptorID":150,"Name":"v4dep","SchemaID":105}}
 │              ├── RemoveDroppedColumnType {"ColumnID":1,"TableID":150}
 │              ├── RemoveDroppedColumnType {"ColumnID":2,"TableID":150}
 │              ├── MarkDescriptorAsDropped {"DescID":149}
 │              ├── RemoveViewBackReferencesInRelations {"BackReferencedViewID":149}
 │              ├── RemoveAllTableComments {"TableID":149}
 │              ├── MarkDescriptorAsDropped {"DescID":150}
 │              ├── RemoveViewBackReferencesInRelations {"BackReferencedViewID":150}
 │              ├── RemoveAllTableComments {"TableID":150}
 │              ├── MarkDescriptorAsDropped {"DescID":148}
 │              ├── RemoveViewBackReferencesInRelations {"BackReferencedViewID":148}
 │              ├── RemoveAllTableComments {"TableID":148}
 │              ├── MarkDescriptorAsDropped {"DescID":147}
 │              ├── RemoveViewBackReferencesInRelations {"BackReferencedViewID":147}
 │              ├── RemoveAllTableComments {"TableID":147}
 │              ├── SetJobStateOnDescriptor {"DescriptorID":146,"Initialize":true}
 │              ├── SetJobStateOnDescriptor {"DescriptorID":147,"Initialize":true}
 │              ├── SetJobStateOnDescriptor {"DescriptorID":148,"Initialize":true}
 │              ├── SetJobStateOnDescriptor {"DescriptorID":149,"Initialize":true}
 │              ├── SetJobStateOnDescriptor {"DescriptorID":150,"Initialize":true}
 │              └── CreateSchemaChangerJob {"NonCancelable":true}
 └── PostCommitNonRevertiblePhase
      └── Stage 1 of 1 in PostCommitNonRevertiblePhase
           ├── 11 elements transitioning toward ABSENT
           │    ├──    DROPPED → ABSENT View:{DescID: 147}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 147, ColumnID: 1}
           │    ├──    DROPPED → ABSENT View:{DescID: 148}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 148, ColumnID: 1}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 148, ColumnID: 2}
           │    ├──    DROPPED → ABSENT View:{DescID: 149}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 149, ColumnID: 1}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 149, ColumnID: 2}
           │    ├──    DROPPED → ABSENT View:{DescID: 150}
           │    ├── WRITE_ONLY → ABSENT Column:{DescID: 150, ColumnID: 1}
           │    └── WRITE_ONLY → ABSENT Column:{DescID: 150, ColumnID: 2}
           └── 21 Mutation operations
                ├── LogEvent {"Statement":"DROP VIEW ‹test›...","StatementTag":"DROP VIEW","TargetStatus":1}
                ├── DeleteDescriptor {"DescriptorID":147}
                ├── MakeColumnAbsent {"ColumnID":1,"TableID":147}
                ├── LogEvent {"Statement":"DROP VIEW ‹test›...","StatementTag":"DROP VIEW","TargetStatus":1}
                ├── DeleteDescriptor {"DescriptorID":148}
                ├── MakeColumnAbsent {"ColumnID":1,"TableID":148}
                ├── MakeColumnAbsent {"ColumnID":2,"TableID":148}
                ├── LogEvent {"Statement":"DROP VIEW ‹test›...","StatementTag":"DROP VIEW","TargetStatus":1}
                ├── DeleteDescriptor {"DescriptorID":149}
                ├── MakeColumnAbsent {"ColumnID":1,"TableID":149}
                ├── MakeColumnAbsent {"ColumnID":2,"TableID":149}
                ├── LogEvent {"Statement":"DROP VIEW ‹test›...","StatementTag":"DROP VIEW","TargetStatus":1}
                ├── DeleteDescriptor {"DescriptorID":150}
                ├── MakeColumnAbsent {"ColumnID":1,"TableID":150}
                ├── MakeColumnAbsent {"ColumnID":2,"TableID":150}
                ├── RemoveJobStateFromDescriptor {"DescriptorID":146}
                ├── RemoveJobStateFromDescriptor {"DescriptorID":147}
                ├── RemoveJobStateFromDescriptor {"DescriptorID":148}
                ├── RemoveJobStateFromDescriptor {"DescriptorID":149}
                ├── RemoveJobStateFromDescriptor {"DescriptorID":150}
                └── UpdateSchemaChangerJob {"IsNonCancelable":true}
```

`EXPLAIN (DDL, VERBOSE)` Example:
```
• Schema change plan for DROP SEQUENCE ‹test›.public.‹sq1› CASCADE;
│
├── • StatementPhase
│   │
│   └── • Stage 1 of 1 in StatementPhase
│       │
│       ├── • 1 element transitioning toward ABSENT
│       │   │
│       │   └── • Sequence:{DescID: 113}
│       │         PUBLIC → TXN_DROPPED
│       │
│       └── • 1 Mutation operation
│           │
│           └── • MarkDescriptorAsDroppedSynthetically
│                 DescID: 113
│
├── • PreCommitPhase
│   │
│   └── • Stage 1 of 1 in PreCommitPhase
│       │
│       ├── • 9 elements transitioning toward ABSENT
│       │   │
│       │   ├── • Namespace:{DescID: 113, Name: sq1, ReferencedDescID: 104}
│       │   │     PUBLIC → ABSENT
│       │   │
│       │   ├── • Owner:{DescID: 113}
│       │   │   │ PUBLIC → ABSENT
│       │   │   │
│       │   │   └── • skip PUBLIC → ABSENT operations
│       │   │         rule: "skip element removal ops on descriptor drop"
│       │   │
│       │   ├── • UserPrivileges:{DescID: 113, Name: admin}
│       │   │   │ PUBLIC → ABSENT
│       │   │   │
│       │   │   └── • skip PUBLIC → ABSENT operations
│       │   │         rule: "skip element removal ops on descriptor drop"
│       │   │
│       │   ├── • UserPrivileges:{DescID: 113, Name: root}
│       │   │   │ PUBLIC → ABSENT
│       │   │   │
│       │   │   └── • skip PUBLIC → ABSENT operations
│       │   │         rule: "skip element removal ops on descriptor drop"
│       │   │
│       │   ├── • Sequence:{DescID: 113}
│       │   │   │ TXN_DROPPED → DROPPED
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT Namespace:{DescID: 113, Name: sq1, ReferencedDescID: 104}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT Owner:{DescID: 113}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT UserPrivileges:{DescID: 113, Name: admin}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT UserPrivileges:{DescID: 113, Name: root}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT ObjectParent:{DescID: 113, ReferencedDescID: 105}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT TableComment:{DescID: 113}
│       │   │   │     rule: "dependent element removal before descriptor drop"
│       │   │   │
│       │   │   ├── • Precedence dependency from ABSENT ColumnDefaultExpression:{DescID: 114, ColumnID: 2}
│       │   │   │     rule: "column DEFAULT removed before dropping dependent types and sequences"
│       │   │   │
│       │   │   └── • Precedence dependency from ABSENT ColumnDefaultExpression:{DescID: 115, ColumnID: 2}
│       │   │         rule: "column DEFAULT removed before dropping dependent types and sequences"
│       │   │
│       │   ├── • ObjectParent:{DescID: 113, ReferencedDescID: 105}
│       │   │     PUBLIC → ABSENT
│       │   │
│       │   ├── • TableComment:{DescID: 113}
│       │   │   │ PUBLIC → ABSENT
│       │   │   │
│       │   │   └── • skip PUBLIC → ABSENT operations
│       │   │         rule: "skip table comment removal ops on descriptor drop"
│       │   │
│       │   ├── • ColumnDefaultExpression:{DescID: 114, ColumnID: 2}
│       │   │     PUBLIC → ABSENT
│       │   │
│       │   └── • ColumnDefaultExpression:{DescID: 115, ColumnID: 2}
│       │         PUBLIC → ABSENT
│       │
│       └── • 11 Mutation operations
│           │
│           ├── • DrainDescriptorName
│           │     Namespace:
│           │       DatabaseID: 104
│           │       DescriptorID: 113
│           │       Name: sq1
│           │       SchemaID: 105
│           │
│           ├── • RemoveColumnDefaultExpression
│           │     ColumnID: 2
│           │     TableID: 114
│           │
│           ├── • UpdateBackReferencesInSequences
│           │     BackReferencedColumnID: 2
│           │     BackReferencedTableID: 114
│           │     SequenceIDs:
│           │     - 113
│           │
│           ├── • RemoveColumnDefaultExpression
│           │     ColumnID: 2
│           │     TableID: 115
│           │
│           ├── • UpdateBackReferencesInSequences
│           │     BackReferencedColumnID: 2
│           │     BackReferencedTableID: 115
│           │     SequenceIDs:
│           │     - 113
│           │
│           ├── • MarkDescriptorAsDropped
│           │     DescID: 113
│           │
│           ├── • RemoveAllTableComments
│           │     TableID: 113
│           │
│           ├── • SetJobStateOnDescriptor
│           │     DescriptorID: 113
│           │     Initialize: true
│           │
│           ├── • SetJobStateOnDescriptor
│           │     DescriptorID: 114
│           │     Initialize: true
│           │
│           ├── • SetJobStateOnDescriptor
│           │     DescriptorID: 115
│           │     Initialize: true
│           │
│           └── • CreateSchemaChangerJob
│                 Authorization:
│                   UserName: root
│                 DescriptorIDs:
│                 - 113
│                 - 114
│                 - 115
│                 JobID: 1
│                 NonCancelable: true
│                 Statements:
│                 - statement: DROP SEQUENCE sq1 CASCADE
│                   redactedstatement: DROP SEQUENCE ‹test›.public.‹sq1› CASCADE
│                   statementtag: DROP SEQUENCE
│
└── • PostCommitNonRevertiblePhase
    │
    └── • Stage 1 of 1 in PostCommitNonRevertiblePhase
        │
        ├── • 1 element transitioning toward ABSENT
        │   │
        │   └── • Sequence:{DescID: 113}
        │         DROPPED → ABSENT
        │
        └── • 6 Mutation operations
            │
            ├── • LogEvent
            │     Authorization:
            │       UserName: root
            │     Element:
            │       Sequence:
            │         sequenceId: 113
            │     Statement: DROP SEQUENCE ‹test›.public.‹sq1› CASCADE
            │     StatementTag: DROP SEQUENCE
            │     TargetMetadata:
            │       SourceElementID: 1
            │       SubWorkID: 1
            │     TargetStatus: 1
            │
            ├── • CreateGcJobForTable
            │     StatementForDropJob:
            │       Statement: DROP SEQUENCE test.public.sq1 CASCADE
            │     TableID: 113
            │
            ├── • RemoveJobStateFromDescriptor
            │     DescriptorID: 113
            │     JobID: 1
            │
            ├── • RemoveJobStateFromDescriptor
            │     DescriptorID: 114
            │     JobID: 1
            │
            ├── • RemoveJobStateFromDescriptor
            │     DescriptorID: 115
            │     JobID: 1
            │
            └── • UpdateSchemaChangerJob
                  IsNonCancelable: true
                  JobID: 1
```

### CODE REVIEW

**Comment by ajwerner** [comment]
One extension (not for this PR) which would be nice, would be to show the other DDL statements from the current transaction before this statement.

Another is that I think this is far more verbose than people want by default. That doesn't need to be addressed in this PR. I think I'd like to see a compact view which says how many stages and for each stage, how many transitions of a given type there are and how many operations of a given type are going to be executed. Maybe showing the full set if it is smaller than something. 

I especially don't think end users want to know anything about rules in their default explain diagram. 

Consider below hacking as some inspiration. I think we just want something really compact that gives a sense of what's going to happen before printing a wall of text. 

```
• Schema change plan for DROP SEQUENCE ‹test›.public.‹sq1› CASCADE;
│
├── • StatementPhase
│   │
│   └── • Stage 1 of 1 in StatementPhase
│       │
│       ├── • 1 element transitioning toward ABSENT
│       │   │
│       │   └── • Sequence:{DescID: 113}
│       │         PUBLIC -> TXN_DROPPED
│       │
│       └── • 1 Mutation operation
│           │
│           └── • MarkDescriptorAsDroppedSynthetically
│                 DescID: 113
│
├── • PreCommitPhase
│   │
│   └── • Stage 1 of 1 in PreCommitPhase
│       │
│       ├── • 9 elements transitioning toward ABSENT
│       │   │
│       │   ├── • Namespace:{DescID: 113, Name: sq1, ReferencedDescID: 104}
│       │   │     PUBLIC -> ABSENT
│       │   │
│       │   ├── • Owner:{DescID: 113}
│       │   │     PUBLIC -> ABSENT
│       │   │
│       │   ├── • UserPrivileges:{DescID: 113, Name: admin}
│       │   │   PUBLIC -> ABSENT
│       │   │
│       │   ├── • UserPrivileges:{DescID: 113, Name: root}
│       │   │   PUBLIC -> ABSENT
│       │   │
│       │   ├── • Sequence:{DescID: 113}
│       │   │     TXN_DROPPED -> DROPPED
│       │   │
│       │   ├── • ObjectParent:{DescID: 113, ReferencedDescID: 105}
│       │   │     PUBLIC -> ABSENT
│       │   │
│       │   ├── • TableComment:{DescID: 113}
│       │   │     PUBLIC -> ABSENT
│       │   │
│       │   ├── • ColumnDefaultExpression:{DescID: 114, ColumnID: 2}
│       │   │     PUBLIC -> ABSENT
│       │   │
│       │   └── • ColumnDefaultExpression:{DescID: 115, ColumnID: 2}
│       │         PUBLIC -> ABSENT
│       │
│       └── • 11 Mutation operations
│           ├── • DrainDescriptorName (113)
│           ├── • RemoveColumnDefaultExpression (114, 2), (115, 2)
│           ├── • UpdateBackReferencesInSequences (114, 2), (115, 2)
│           ├── • MarkDescriptorAsDropped (113)
│           ├── • RemoveAllTableComments (113)
│           ├── • SetJobStateOnDescriptor (113, 114, 115)
│           └── • CreateSchemaChangerJob
│
└── • PostCommitNonRevertiblePhase
    │
    └── • Stage 1 of 1 in PostCommitNonRevertiblePhase
        │
        ├── • 1 element transitioning toward ABSENT
        │   │
        │   └── • Sequence:{DescID: 113}
        │         DROPPED -> ABSENT
        │
        └── • 6 Mutation operations
            │
            ├── • LogEvent (DROP SEQUENCE)
            ├── • CreateGcJobForTable (113)
            ├── • RemoveJobStateFromDescriptor (113), (114), (115)
            └── • UpdateSchemaChangerJob
```

**Comment by postamar** [comment]
I like your example of something more compact. I think that in this PR I'll have `EXPLAIN (DDL)` do something like that by default and print the rules if a flag is set (DEPS?). Shouldn't be hard to do.

**Comment by ajwerner** [comment]
>I like your example of something more compact. I think that in this PR I'll have EXPLAIN (DDL) do something like that by default and print the rules if a flag is set (DEPS?). Shouldn't be hard to do.

That sounds good to me. The term `VERBOSE` is used in other explains as an option. I think it'd be okay to just have two modes, normal and verbose.

**Comment by postamar** [comment]
I have done just that, RFAL.

**Comment by ajwerner** [comment]
I think you should avoid printing out the ops we skipped in the compact form. It's confusing to show them. More or less same feeling about transitions related to elided op edges. 

Also, thoughts on making the state transitions into a single line? Ideally aligned such that you use a fixed width for the start and end state. 

```
 │         │    ├── Column:{DescID: 149, ColumnID: 1}
 │         │    │   PUBLIC -> WRITE_ONLY
 │         │    ├── ColumnName:{DescID: 149, Name: name, ColumnID: 1}
 │         │    │   PUBLIC -> ABSENT
````
to
```
 │         │    ├── PUBLIC -> WRITE_ONLY Column:{DescID: 149, ColumnID: 1}
 │         │    ├── PUBLIC -> ABSENT     ColumnName:{DescID: 149, Name: name, ColumnID: 1}
 ```

**Comment by postamar** [comment]
> I think you should avoid printing out the ops we skipped in the compact form. It's confusing to show them.

Yikes that wasn't intentional. 

> More or less same feeling about transitions related to elided op edges.

I don't agree with you there, I think all targets should appear in the plan.

> Also, thoughts on making the state transitions into a single line?  Ideally aligned such that you use a fixed width for the start and end state.

Good idea.

**Comment by postamar** [comment]
Regarding the skipped ops perhaps you were misled by the `RemoveDroppedColumnType` and `MakeColumnAbsent` ops in the `DROP VIEW CASCADE`. These really should be marked as no-op, but aren't. They don't do anything if the descriptor is already marked as dropped.

They're there because they're bound to an op edge which also emit operations for removing back-references. Of course in these cases as the columns don't have a user-defined type that operation doesn't get emitted.

tl;dr we need more expressive rules

**Comment by postamar** [comment]
I updated the PR description with the result of my latest changes. Width-alignment is nice. Thanks for that.

**Review by ajwerner — APPROVED** [review]
LGTM! Thanks for doing this!

File: pkg/sql/schemachanger/scplan/plan_explain.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+					depEdgeByElement[t.Element()] = append(depEdgeByElement[t.Element()], de)
+					return nil
+				}); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	// Generate format string for printing element status transition.
+	fmtVerboseTransition := fmt.Sprintf("%%%ds → %%-%ds %%s", beforeMaxLen, afterMaxLen)
```
[the remark this answers is not in the document]
Comment (postamar) [reply pkg/sql/schemachanger/scplan/plan_explain.go:178]: Good idea.

**Comment by postamar** [comment]
Thanks for the review and thanks for the helpful suggestions!

bors r+
